### ISSUE TITLE
Dependent Filter Issue :- If the parent Filter value is changed then the dependent filter with  previous selected value is not getting removed if its not in the dependent filter data(options) 

### ISSUE THREAD

**Issue by shubham1811** [apache/superset#24511]
How to reproduce the bug

1. Create a Filter Name **A**
2. Create a Filter Name **B** 
3. Tick the checkbox the **Values are dependent on other filters**  in Filter **B** and select **A**  in **VALUES DEPENDENT ON**
4. Click on **Save** 
5. Select some value in filter **A**
6. Data API call will be done for Filter **B** and the option will be updated. 
7. Select some value in Filter **B**. For Example, Selected **Asia**. 
8. Remove the Filter Value in Filter **A** and select some other option. 
9. Data API call will be done for Filter **B** and the option will be updated. 
10. In Option of Filter **B** previously selected will be present and it's not coming from the Result of Data(API Call). 

### Expected results

If The Previously selected value is not present in the result of the dependent filter API call it shouldn't be present in the option or in the selected state. 

### Actual results

The previously selected value if it's not present in the result of the dependent filter API call its showing in the option or its getting selected also. 

#### Screenshots

[https://jam.dev/c/2e22ee9e-d1b2-48ba-a063-11dfe2584c8a](url)
<img width="707" alt="Screenshot 2023-06-26 at 12 32 57 PM" src="https://github.com/apache/superset/assets/26042498/d883cc4d-6398-4274-a75b-1d9c0af31a88">

[https://jam.dev/c/54743133-2a1e-4d79-bfa4-30e0d8cdc95c](url)
<img width="703" alt="Screenshot 2023-06-26 at 12 36 28 PM" src="https://github.com/apache/superset/assets/26042498/59f23b89-2321-4259-927b-e590d21367e2">

[https://jam.dev/c/bad93252-6c4c-4ade-bd7b-53108b245692](url)

<img width="1427" alt="Screenshot 2023-06-26 at 12 40 19 PM" src="https://github.com/apache/superset/assets/26042498/9cf32efb-d06c-46b2-bf3f-96976141b79c">

[https://jam.dev/c/2e343713-d305-4b3d-b504-8c5d48b75647](url)
<img width="820" alt="Screenshot 2023-06-26 at 12 41 34 PM" src="https://github.com/apache/superset/assets/26042498/a3698a39-cbb8-4708-93d1-19eabc8c1526">

[https://jam.dev/c/5f3fa04b-c25f-4ea9-9108-c1059a3f9caf](url)
<img width="1431" alt="Screenshot 2023-06-26 at 12 44 10 PM" src="https://github.com/apache/superset/assets/26042498/0b00cd62-122d-46a6-bda0-4f72056d83cb">

https://jam.dev/c/c963fc06-e234-422e-85f5-60a8b7592574
<img width="815" alt="Screenshot 2023-06-26 at 12 46 59 PM" src="https://github.com/apache/superset/assets/26042498/3d1b878d-4b3e-45e7-bab2-5e3efcdb8520">

### Environment

(please complete the following information):

- browser type and version:
- superset version: `master `
- python version: `python --version`
- node.js version: `16.15.0`
- any feature flags active:

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [Yes] I have checked the superset logs for Python stack traces and included them here as text if there are any.
- [Yes] I have reproduced the issue with at least the latest released version of the superset.
- [Yes] I have checked the issue tracker for the same issue and I haven't found one similar.

**Comment by sfirke** [comment]
I can validate this in Superset 2.1.0 and agree that it would be better if when the user selects a new value in filter A, the value in filter B is reset.  I thought this issue had already been filed but I'm not seeing another bug report so thanks for submitting this one.

**Comment by lchavas** [comment]
Having the same issue here.

I can add that :
- if you have 'Select first filter value by default ' active on the filters and click on the 'CLEAR ALL' button while parent filter is loading, you'll get 'null' as a selected value in child filters
- if you have 'Select first filter value by default ' active on the filters and then click on the 'CLEAR ALL' button while parent and child filters are set, it will reset the parent filter but not the children. another 'CLEAR ALL' will then clear the children.

It feels like any change in the parent should cascade in the child filter, and child filters should be loaded only when the parent filter has finished loading.

**Comment by shubham1811** [comment]
> Having the same issue here in 2.1.0
> 
> I can add that :
> 
> * if you have 'Select first filter value by default ' active on the filters and click on the 'CLEAR ALL' button while parent filter is loading, you'll get 'null' as a selected value in child filters
> * if you have 'Select first filter value by default ' active on the filters and then click on the 'CLEAR ALL' button while parent and child filters are set, it will reset the parent filter but not the children. another 'CLEAR ALL' will then clear the children.
> 
> It feels like any change in the parent should cascade in the child filter, and child filters should be loaded only when the parent filter has finished loading.

Yes Correct child filters should be loaded only when the parent filter has finished loading.

**Comment by lchavas** [comment]
Updated to 2.1.1 with high hopes, parent filter being adressed in the changelog.
However the issue is unchanged, and we still have to clear the first null value manually and change the subfilters manually everytime the parent filter is changed.

**Comment by Lbostwick** [comment]
Does anyone know if this is addressed in SuperSet's latest update to version 3.0?

**Comment by Ian2012** [comment]
This is still happening in superset 3.0

**Comment by giacomochiarella** [comment]
@rusackas I can confirm it is still happening in 4.0.0 also

**Comment by hk-the-analyst** [comment]
Facing the same issue in 10.1.1

**Comment by cmartinezdnn** [comment]
I can confirm it is not fixed in 4.0.2

**Comment by cmartinezdnn** [comment]
This issue is very annoying. Does anybody know how to fix it?

**Comment by hk-the-analyst** [comment]
Still facing this issue in 4.1.1

**Comment by rusackas** [comment]
Sorry to see this hasn't been prioritized... perhaps since it's a bit of an edge case. Investigation/PRs welcome. 

Perhaps @dosu can shed some light on this.

**Comment by rusackas** [comment]
This is a related/duplicate of #22979 (dependent filters not clearing when parent changes). The bug is still present in recent versions. A fix would involve clearing or re-filtering the dependent filter's selected value when the parent filter changes and the previously-selected value is no longer valid. PRs welcome.

**Comment by rusackas** [comment]
https://github.com/apache/superset/pull/39504 seems to have fixed at least PART of this ancient issue. Anyone able to tell if we should keep this open? Personally, I'd love to close this one out, and file an updated issue with whatever work is left here.

**Comment by rusackas** [comment]
Executive decision time... I'm going to close this as fixed. If there's still an issue remaining after the merged PR, let's open a fresh issue with updated context/repro
